### PR TITLE
GDScript: Fix doc comment parsing (newlines and spaces)

### LINKED ISSUE #76815: Multi-line class documentation comments begin with an extra space
### Godot version

v4.0.stable.official [92bee43ad]

### System information

Steam OS Holo, Steam Deck

### Issue description

Using a `[br]` tag at the end of a line in a documentation comment adds an additional space at the beginning of the following line:

image

### Steps to reproduce

The script below is what I used to generate the image above:

```gdscript
class_name Test
extends Node

## High-level description
##
## First line of detail.[br]
## Second line of detail.
```

Removing the space between `##` and `Second` doesn't change the output.

The only workaround I've found so far is to remove the linebreak after the `[br]` tag, but this makes the documentation comments harder to read in the editor:

```gdscript
class_name Test
extends Node

## High-level description
##
## First line of detail.[br]Second line of detail.
```

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by dalexeev**
* Do not add an extra space after `[br]`.
* Do not add an extra space at the end of codeblock.
* Fix text overlapping after `[/codeblock]`.
* More correct processing of codeblocks.
* `@tutorial` and blank line no longer resets `in_codeblock` in class doc comment.

```gdscript
## Line
## Continue[br]
## New Line
## [codeblock]
## if true:
##     print(42)
## [/codeblock]
## Test
func test():
    pass
```

Before:

After:

Another test:

```gdscript
## Line
## Continue[br]
## New Line[br]
## Test[br]Test
## [codeblock]
## test
## test[br]
## test[br]test
##
## @tutorial
## [/codeblock]
## test
```

Closes #76815.

### CODE REVIEW

File: modules/gdscript/gdscript_parser.cpp [hunk 1]
```diff
@@ -3261,31 +3261,85 @@ GDScriptParser::TypeNode *GDScriptParser::parse_type(bool p_allow_void) {
...
+			if (line[i] != ' ') {
+				break;
+			}
+		}
+		// ## [codeblock]
+		// ## if true:
+		// ##^
+		// ##     print(42)
+		// ##^^^^^
+		line = line.substr(i % 4);
```
Comment (dalexeev) [inline modules/gdscript/gdscript_parser.cpp:3280]: I'm not sure how best to handle this.

Comment (MewPurPur) [reply modules/gdscript/gdscript_parser.cpp:3280]: Why not use lstrip?

Comment (dalexeev) [reply modules/gdscript/gdscript_parser.cpp:3280]: We need to keep the indentation inside the codeblock. Ideally, all of these options

```gdscript
##[codeblock]
##if true:
##    print(42)
```

```gdscript
## [codeblock]
## if true:
##     print(42)
```

```gdscript
##  [codeblock]
##  if true:
##      print(42)
```

should be converted to

```
[codeblock]
if true:
    print(42)
```

**Comment by akien-mga** [comment]
CC @anvilfolk

File: modules/gdscript/gdscript_parser.cpp (hunk 1 above)
Comment (dalexeev) [reply modules/gdscript/gdscript_parser.cpp]: Done.

**Comment by dalexeev** [comment]
The main problem that this PR solves is the correct detection of line join, depending on whether the current line is inside `[codeblock][/codeblock]` or not, and also whether the previous line ends with `[br]` or not.

* Outside codeblock:
  * If the text is empty, line join is empty.
  * If the previous line ends with `[br]`, line join is empty.
  * Otherwise line join is one space.
  * Extra spaces must be removed.
* Inside codeblock:
  * All spaces and newlines must be kept as is (except for the initial indent after `##`).

For example, the following comment

```gdscript
## Line
## Continue[br]
## New Line[br]
## Test[br]Test
##   Test
## [codeblock]
## test
## test[br]
##     test[br]test
## [/codeblock]
## test
```

should be converted to

```
Line Continue[br]New Line[br]Test[br]Test Test
[codeblock]
test
test[br]
    test[br]test
[/codeblock]
test
```

Also, although this is not a recommended syntax, the following comment

```gdscript
## aaa[codeblock]bbb[/codeblock]ccc[codeblock]ddd[/codeblock]eee
```

should be converted to

```
aaa
[codeblock]
bbb
[/codeblock]
ccc
[codeblock]
ddd
[/codeblock]
eee
```

I recommend "Hide whitespace" to make diff easier.

**Comment by dalexeev** [comment]
Added check for `[code][/code]` (for example `[code][codeblock][/code]`).

**Review by akien-mga — APPROVED** [review]
LGTM, though TIWAGOS. I haven't reviewed in depth but it's also touching a non-critical part of the parser (only affects editor behavior) so it seems fairly safe to me to merge.

File: modules/gdscript/gdscript_parser.cpp [hunk 2]
```diff
@@ -3293,31 +3293,116 @@ GDScriptParser::TypeNode *GDScriptParser::parse_type(bool p_allow_void) {
...
-	if (start_block != -1 && r_block_begins) {
-		*r_block_begins = start_block;
+enum DocLineState {
+	DOC_LINE_NORMAL,
+	DOC_LINE_IN_CODE,
+	DOC_LINE_IN_CODEBLOCK,
+};
+
+static String _process_doc_line(const String &p_line, const String &p_text, const String &p_space_prefix, DocLineState *r_state) {
+	ERR_FAIL_COND_V(r_state == nullptr, String());
```
Comment (anvilfolk) [inline modules/gdscript/gdscript_parser.cpp:3303]: Could you make `r_state` a reference if it's never meant to be `null`?

Comment (dalexeev) [reply modules/gdscript/gdscript_parser.cpp]: Done.

File: modules/gdscript/gdscript_parser.cpp [hunk 3]
```diff
@@ -3293,31 +3293,116 @@ GDScriptParser::TypeNode *GDScriptParser::parse_type(bool p_allow_void) {
...
+					process = false;
+					break;
+				}
+
+				from = rb_pos + 1;
+
+				String tag = line.substr(lb_pos + 1, rb_pos - lb_pos - 1);
+				if (tag == "code") {
+					*r_state = DOC_LINE_IN_CODE;
+				} else if (tag == "codeblock") {
```
[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/gdscript_parser.cpp:3347]: Makes sense :)

**Comment by anvilfolk** [comment]
> > For instance, I would love to know ta a glance what the difference is between `DOC_LINE_IN_CODE` and `DOC_LINE_IN_CODE_BLOCK`. I assume it's the back-quote vs the actual `[codeblock]`?, thus allowing back-quotes to have multiline strings?
> 
> `DOC_LINE_IN_CODE` - in `[code][/code]` `DOC_LINE_IN_CODEBLOCK` - in `[codeblock][/codeblock]`

You're right, once I finally got through parsing some of the code it wasn't so hard to find :)

**Pull request merged**